tools/ci_connector_ops/ci_connector_ops/pipelines/utils.py

```diff
@@ -73,57 +110,78 @@ async def get_file_contents(container: Container, path: str) -> Optional[str]:
-            # this is the hicky bit of the stopgap because
             # this error could come from a network issue
             raise
     return None
 
 
-# This is a stop-gap solution to capture non 0 exit code on Containers
-# The original issue is tracked here https://github.com/dagger/dagger/issues/3192
+async def get_container_output(container: Container) -> Tuple[str, str]:
+    """Retrieve both stdout and stderr of a container, concurrently.
+
+    Args:
+        container (Container): The container to execute.
+
+    Returns:
+        Tuple[str, str]: The stdout and stderr of the container, respectively.
+
+    Raises:
+        ExecError: If the container exit code is not 0.
+    """
+    async with asyncer.create_task_group() as task_group:
+        soon_stdout = task_group.soonify(container.stdout)
+        soon_stderr = task_group.soonify(container.stderr)
+    return soon_stdout.value, soon_stderr.value
+
+
+async def get_exec_result(container: Container) -> Tuple[int, str, str]:
+    """Retrieve the exit_code along with stdout and stderr of a container by handling the ExecError.
+
+    Note: It is preferrable to not worry about the exit code value and just capture
+    ExecError to handle errors. This is offered as a convenience when the exit code
+    value is actually needed.
+
+    Args:
+        container (Container): The container to execute.
+
+    Returns:
+        Tuple[int, str, str]: The exit_code, stdout and stderr of the container, respectively.
+    """
+    try:
+        return 0, *(await get_container_output(container))
+    except ExecError as e:
+        return e.exit_code, e.stdout, e.stderr
+
+
 async def with_exit_code(container: Container) -> int:
     """Read the container exit code.
 
-    If the exit code is not 0 a QueryError is raised. We extract the non-zero exit code from the QueryError message.
-
     Args:
         container (Container): The container from which you want to read the exit code.
 
     Returns:
         int: The exit code.
     """
     try:
-        await container.exit_code()
-    except QueryError as e:
-        error_message = str(e)
-        if "exit code: " in error_message:
-            exit_code = re.search(r"exit code: (\d+)", error_message)
-            if exit_code:
-                return int(exit_code.group(1))
-            else:
-                return 1
-        raise
+        await container
+    except ExecError as e:
+        return e.exit_code
     return 0
 
 
-# This is a stop-gap solution to capture non 0 exit code on Containers
-# The original issue is tracked here https://github.com/dagger/dagger/issues/3192
 async def with_stderr(container: Container) -> str:
-    """Retrieve the stderr of a container and handle unexpected errors."""
+    """Retrieve the stderr of a container even on execution error."""
     try:
         return await container.stderr()
-    except QueryError as e:
-        return str(e)
+    except ExecError as e:
+        return e.stderr
```

Comment:
Before 0.6.0, the error message from the API would include `Stdout:\n{stdout}` and `Stderr:\n{stderr}` lines. This was removed in the API from the error message into extension fields, which are only read by `ExecError` now. 

Additionally:
- `str(e)` on a `QueryError` would previously return the whole error message, exit code, stdout and stderr. Now it only returns the error message.
- As a convenience, if you do `str(e)` on an `ExecError` you'll get the same output as before (i.e., included "Stdout:" and "Stderr:" lines).

**Note** that I've changed from `str(e)` to `e.stderr` since that seems to be the original intent. Otherwise, when you do this:
```python
stdout = await with_stdout(ctr)
stderr = await with_stderr(ctr)
return StepResult(
    ...
    stdout=stdout,
    stderr=stderr,
)
```
You'd get the same output in both `stdout` and `stderr` instead of just what you intend (in case of error).

**However**, now you don't have the **error message** if you just return `e.stderr`. If you need it in the result for the report, I suggest adding a field for it in `StepResult`, or prepending `e.message` to `e.stderr`, whatever makes the most sense for you at the moment.

If you want to return to the old behavior, just replace `QueryError` with `ExecError` and keep `str(e)`:
```python
try:
    return await container.stderr()
except ExecError as e:
    return str(e)
```